packages/odyssey-react/src/components/TextInput/TextInput.tsx

```diff
@@ -145,28 +147,62 @@ export const TextInput = withTheme(
       ...rest
     } = props;
 
-    const isSearchTextInput = type === "search";
-
     const oid = useOid(id);
     const omitProps = useOmit(rest);
     const internalRef = useSharedRef(ref);
+    const [isControlled, setIsControlled] = useState(
+      typeof value !== "undefined"
+    );
+    const [hasUncontrolledValue, setHasUncontrolledValue] =
+      useState(defaultValue);
     const [isValid, setIsValid] = useState(checkInputValidity(internalRef));
+    const [hasFocus, setHasFocus] = useState(false);
 
     useEffect(() => {
       setIsValid(checkInputValidity(internalRef));
-    }, [internalRef, required, type, value]);
+      if (!isControlled && typeof value !== "undefined") {
+        setIsControlled(true);
+      }
+    }, [internalRef, isControlled, required, type, value]);
 
     const setFocus = () => {
       requestAnimationFrame(() => {
         internalRef.current?.focus();
       });
     };
 
+    const onClear = () => {
+      setFocus();
+      if (internalRef.current) {
+        const nativeInputValueSetter = Object.getOwnPropertyDescriptor(
+          window?.HTMLInputElement?.prototype,
+          "value"
+        )?.set;
+        if (nativeInputValueSetter) {
+          nativeInputValueSetter.call(internalRef.current, "");
+          const aChangeEvent = new Event("change", { bubbles: true });
+          internalRef.current.dispatchEvent(aChangeEvent);
+        }
```

Comment:
hmmm we still fire the event even for controlled inputs where the value will not change? would be nice to avoid the imperative DOM code if possible!

Reply:
The onChange is always fired for the controlled case when the user types, so I think this consistent?

Reply:
oh interesting - I can't imagine there being alot of utility in adding change event handler callbacks for controlled components but perhaps there is nuance I'm not aware of! I think I would prefer to keep this simple and not show the clear button or fire the event for controlled components! The accessibility concerns of this are something we can improve in the future with better documentation around adding supporting context to the input label etc.

Reply:
This was written from my perspective, as a user of this component, if I was controlling the value, I would expect the `onChange` to be raised on any change and I would expect the clear button to show on hover and focus, and I would expect clicking the button to raise an `onChange` with an empty value.
But I don't have much insight into the accessibility concerns